presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java

```diff
@@ -58,62 +52,133 @@ public Block distinct(@TypeParameter("E") Type type, @SqlType("array(E)") Block
             }
         }
 
-        TypedSet typedSet = new TypedSet(type, array.getPositionCount(), "array_distinct");
-        int distinctCount = 0;
+        TypedSet typedSet = new TypedSet(type, arrayLength, "array_distinct");
+        BlockBuilder distinctElementBlockBuilder;
+        int position = 0;
 
-        if (pageBuilder.isFull()) {
-            pageBuilder.reset();
-        }
+        if (array.mayHaveNull()) {
+            while (position < arrayLength && typedSet.add(array, position)) {
+                position++;
+            }
 
-        BlockBuilder distinctElementBlockBuilder = pageBuilder.getBlockBuilder(0);
-        for (int i = 0; i < array.getPositionCount(); i++) {
-            if (!typedSet.contains(array, i)) {
-                typedSet.add(array, i);
-                distinctCount++;
+            if (position == arrayLength) {
+                // All elements are distinct, so just return the original.
+                return array;
+            }
+
+            distinctElementBlockBuilder = type.createBlockBuilder(null, arrayLength);
```

Comment:
Wow. What's the benefit of having 59-68? -- short circuit when array elements are already distinct?  Maybe add some comments?

Reply:
ping @kaikalur

Reply:
Done. Added comments.